include/envoy/config/typed_config.h

```diff
@@ -0,0 +1,37 @@
+#pragma once
+
+#include "envoy/common/pure.h"
+
+#include "common/protobuf/protobuf.h"
+
+namespace Envoy {
+namespace Config {
+
+/**
+ * Base class for config defined by a typed proto message.
+ */
+class TypedConfig {
+public:
+  /**
+   * @return ProtobufTypes::MessagePtr create empty config proto message for v2. The config, which
+   * arrives in an opaque google.protobuf.Struct message, will be converted to JSON and then parsed
+   * into this empty proto.
+   */
+  virtual ProtobufTypes::MessagePtr createEmptyConfigProto() PURE;
+
+  /**
+   * @return config proto full name
+   */
+  virtual std::string type() {
```

Comment:
Would it make sense to name it `config_type_url()`?

Reply:
Maybe `config_type`? I didn't want to call it URL since it's not really a type URL as in Any. Calling message name is also confusing (unless you're used to protobuf lingua).

Reply:
If we're going to bike shed, what about we replace the whole category and config type thing with `kind`, as in https://en.wikipedia.org/wiki/Kind_(type_theory).

Reply:
Kind might confused k8s user, since in k8s the API group and version are separated from the kind. Kind means an "object kind" in that object-oriented world.

Reply:
Why not use the type URL as in Any?

Reply:
Type URL is ambiguous. It looks like `proto_registry/proto_name` where `proto_registry is usually `type.googleapis.com` but it doesn't have to be. It's some sort of indicator about where to fetch the descriptor, although I haven't seen anyone using it that way.

Reply:
Ok, if proto descriptor is less ambiguous than type URL than I should change corresponding field name in the  Extension proto in PR #9301